cluster/assemblestate/assemblestate.go

```diff
@@ -0,0 +1,880 @@
+	validated, err := validateSession(session)
+	if err != nil {
+		return nil, fmt.Errorf("invalid session data: %w", err)
+	}
+
+	if log == nil {
+		log = logger.NullLogger
+	}
+
+	cert, err := tls.X509KeyPair([]byte(config.TLSCert), []byte(config.TLSKey))
+	if err != nil {
+		return nil, err
+	}
+
+	if err := ensureLocalDevicePresent(&session.Devices, Identity{
+		RDT: config.RDT,
+		FP:  CalculateFP(cert.Certificate[0]),
+	}); err != nil {
+		return nil, err
+	}
+
+	devices := NewDeviceQueryTracker(session.Devices, time.Minute*5, config.Clock)
+
+	sel, err := selector(config.RDT, devices.Identified)
+	if err != nil {
+		return nil, err
+	}
+
+	// inform the selector of any routes that we already know. we state that
+	// their provenance is this local node, since we don't persist which routes
+	// came from which peer. this will lead to our local node doing some wasted
+	// publications, but that is okay.
+	if _, _, err := sel.RecordRoutes(config.RDT, session.Routes); err != nil {
+		return nil, err
+	}
+
+	// for any peers that we already trust and know their address, we can safely
+	// inform the selector that the route from our local node to that peer can
+	// be published
+	for fp, peer := range validated.trusted {
+		addr, ok := validated.addresses[fp]
+		if !ok {
+			continue
+		}
+
+		sel.AddAuthoritativeRoute(peer.RDT, addr)
+	}
+
+	// calculate the HMAC once for this device's authentication
+	authHMAC := CalculateHMAC(config.RDT, CalculateFP(cert.Certificate[0]), config.Secret)
+
+	as := AssembleState{
+		initiated:    session.Initiated,
+		config:       config,
+		logger:       log,
+		commit:       commit,
+		clock:        config.Clock,
+		cert:         cert,
+		authHMAC:     authHMAC,
+		trusted:      validated.trusted,
+		fingerprints: validated.fingerprints,
+		addresses:    validated.addresses,
+		discovered:   validated.discovered,
+		devices:      devices,
+		selector:     sel,
+	}
+
+	return &as, nil
+}
+
+type AssembleConfig struct {
+	Secret  string           `json:"secret"`
+	RDT     DeviceToken      `json:"rdt"`
+	IP      net.IP           `json:"ip"`
+	Port    int              `json:"port"`
+	TLSCert []byte           `json:"cert"`
+	TLSKey  []byte           `json:"key"`
+	Clock   func() time.Time `json:"-"`
+}
+
+// publishAuth uses the given [Client] to publish to each device. If publication
+// succeeds, then the certificate returned by [Client.Untrusted] (the
+// certificate that the peer used to communicate with us) is associated with the
+// address that we reached them at.
+//
+// If the certificate the peer used is already trusted (they've already
+// published their auth message to us), then we verify the route from this local
+// node to that peer.
+func (as *AssembleState) publishAuth(ctx context.Context, addresses []string, client Client) error {
+	as.lock.Lock()
+	defer as.lock.Unlock()
+
+	for _, addr := range addresses {
+		if as.discovered[addr] {
+			continue
+		}
+
+		as.lock.Unlock()
```

Comment:
we probably want to wrap to be able to use defers around anything that does something not super simple